### PR TITLE
Feature/order routes

### LINKED ISSUE #62: possibility to order the registrations
Since express has the possibility to add default routes (routes with `*` in it) we need to provide any way of ordering the registrations.

If we don't, the (random) order of the decorators would define the wildcard routes.

One usecase:
- api controller with `/api/....`
- static file controller for a single page application `/*` with a catch all that is not registered to send the `index.html`

### PR DESCRIPTION
**Pull request by buehler**
Closes #62 

Orders all routes by the following criteria:
- The more segments (split with `/`), the earlier are they registered.
- For each segment-count: the less wildcards (`*`), the earlier registered.

### CODE REVIEW

File: controllers/ControllerDecorator.ts [hunk 1]
```diff
@@ -49,6 +49,23 @@ class ControllerRegistration {
...
+        this.wildcardCount = this.routeUrl.split('*').length - 1;
+        this.segmentCount = this.routeUrl.split('/').length;
+    }
+}
+
+function wildcardSort(o1: RegistrationHelper, o2: RegistrationHelper): number {
+    if (o1.wildcardCount === o2.wildcardCount) {
+        return 0;
+    }
+    return o1.wildcardCount > o2.wildcardCount ? 1 : -1;
```
Comment (dwirz) [inline controllers/ControllerDecorator.ts]: short version `return o1.wildcardCount - o2.wildcardCount;` or as lambda function?

File: controllers/ControllerDecorator.ts [hunk 2]
```diff
@@ -292,10 +309,24 @@ export function registerControllers(baseUrl: string = '', router: Router = Route
...
+        }
+
+        map[route.segmentCount].push(route);
+    }
+
+    for (let routes of map.filter(Boolean).reverse()) {
+        for (let route of routes.sort(wildcardSort)) {
+            registerRoute(route, router);
+        }
+    }
```
Comment (dwirz) [inline controllers/ControllerDecorator.ts]: remembered that i had a similar sort problem... what about this: 

``` typescript
definedRoutes
  .reduce((segmentSorted, r) => {
    (segmentSorted[r.segmentCount] || (segmentSorted[r.segmentCount] = [])).push(r);
    return segmentSorted;
  }, [])
  .filter(Boolean)
  .reverse()
  .reduce((a, b) => a.concat(b.sort((r1, r2) => r1.wildcardCount - r2.wildcardCount)), []) // lambda for wildcardSort or use wildcardSort
  .forEach(r => registerRoute(r, router));
```

Comment (buehler) [inline controllers/ControllerDecorator.ts]: 😮

File: controllers/ControllerDecorator.ts (hunk 1 above)
Comment (buehler) [inline controllers/ControllerDecorator.ts]: oh yea, that's the javascript way of compare-functions :-) I did the `C#` variant

File: controllers/ControllerDecorator.ts (hunk 2 above)
Comment (buehler) [inline controllers/ControllerDecorator.ts]: 17899d255f715b2af19ce17cfe812d086f482f44

File: controllers/ControllerDecorator.ts (hunk 1 above)
Comment (buehler) [inline controllers/ControllerDecorator.ts]: 17899d255f715b2af19ce17cfe812d086f482f44

**Pull request merged**